File: index/index-list-iterative-pagination.js
import { TwelveLabs } from 'twelvelabs-js';

function printIndexPage(indexPage) {
  indexPage.forEach(index => {
    console.log(`ID: ${index.id}`);
    console.log(`  Name: ${index.name}`);
    console.log("  Engines:");
    index.engines.forEach((engine, index) => {
      console.log(`    Engine ${index + 1}:`);
      console.log(`      Name: ${engine.name}`);
      console.log(`      Options: ${JSON.stringify(engine.options)}`);
    });
    console.log(`  Video count: ${index.videoCount}`);
    console.log(`  Total duration: ${index.totalDuration} seconds`);
    console.log(`  Created at: ${index.createdAt}`);
    if (index.updatedAt) {
      console.log(`  Updated at: ${index.updatedAt}`);
    }
  });
}


(async () => {
  const client = new TwelveLabs({ apiKey: process.env.API_KEY}); 

  const listParams = {
    id: '',
    name: '',
    engineOptions: ['conversation', 'visual'],
    engineFamily: 'marengo',
    page: 1,
    pageLimit: 5,
    sortBy: 'updated_at',
    sortOption: 'asc'
  };

  // Fetch the initial page of results
  const indexPaginator = await client.index.listPagination(listParams);

  // Print the first page of results
  printIndexPage(indexPaginator.data);

  // Iterate through subsequent pages
  while (true) {
    const nextIndexPage = await indexPaginator.next();
    if (!nextIndexPage) {
      console.log('No more pages of index results available');
      break;
    }
    printIndexPage(nextIndexPage);
  }
})();

File: index/index-list-iterative-pagination.py
from twelvelabs import TwelveLabs
import os

client = TwelveLabs(api_key=os.environ.get("API_KEY"))

def print_page(page):
    for index in page:
      print(f"ID: {index.id}")
      print(f"  Name: {index.name}")
      print("  Engines:")
      for i, engine in enumerate(index.engines, 1):
          print(f"    Engine {i}:")
          print(f"      Name: {engine.name}")
          print(f"      Options: {engine.options}")
      print(f"  Video count: {index.video_count}")
      print(f"  Total duration: {index.total_duration} seconds")
      print(f"  Created at: {index.created_at}")
      if index.updated_at:
          print(f"  Updated at: {index.updated_at}")

# Fetch the initial page of results
index_paginator = client.index.list_pagination(
    id="<YOUR_INDEX_ID>",
    name="<YOUR_INDEX_NAME>",
    page=1,
    page_limit=5,
    engine_options=["visual", "conversation"],
    engine_family="marengo",
    sort_by = "updated_at",
    sort_option="asc",
    created_at="2024-09-17T07:53:46.365Z",
    updated_at="2024-09-17T07:53:46.365Z"
)

# Print the first page of results
print_page(index_paginator.data)

# Iterate through subsequent pages
while True:
    try:
        next_index_page = next(index_paginator)
        print_page(next_index_page)
    except StopIteration:
        break
    
